### ISSUE TITLE
revisit usecase for ATTRS attribute in hooks (overriding Python module attributes in frozen state)

### ISSUE THREAD

**Issue by matysek** [pyinstaller/pyinstaller#1592]
Import hook attribute `attrs` is already broken for several years and some hook files (hook-xml.*.py) use it. These hooks are thus broken.

We should:
* decide if we want to support global `attrs` in import hooks
* find proper usecase for that:

  * this use case overlaps with rthooks:
    * It looks like it  should allow to override some module attributes for frozen state with a fixed value
    * some rthooks also overrided some module attributes but rthoos allow to override module attributes with dynamic values not  only fixed.
* implementation:
  * Implement it just as `attrs` global hook attribute or
  * as API for any `def hook()` function?

Comments?

**Comment by htgoebel** [comment]
According to [the manual](https://github.com/pyinstaller/pyinstaller/blob/v2.1/doc/source/Manual.rst#hooks-in-detail):
> The main purpose is so that ImportTracker will not issue spurious warnings when the rightmost node in a dotted name turns out to be an attribute in a package, instead of a missing submodule. For an example see the hook file `hook-xml.sax.py.`

modulegraph has a somewhat similar but automated approach by collecting assignments to global variables. But *I assume* at the moment this does not work as expected due to #1550. So IMHO we should fix this in modulegraph and completely drop the `attrs` in hooks.

To fix collecting assignments to global variables (assuming modulegraph did it right prior to our change) we need to create single list containing both imports and globals so we have them in order.

**Comment by htgoebel** [comment]
I'm putting this on my todo-list, since I've made the proposal how I would fix it.

**Comment by htgoebel** [comment]
I checked this again. Conslusion:

1. **`attrs` can competently be removed**
2. No change to modulegraph necessary - my analysis in https://github.com/pyinstaller/pyinstaller/issues/1592#issuecomment-147230772 was wrong
3. Le'ts remove `attrs`.

**Re 2:** first :-)

The hooks `attrs` adds names to the module after it has been imported. This is what modulegraph basically does, too. (Modulegraph stores them in a separate list `globalnames` and "knowns" about them, but this does not matter.) Both ways effect the modules imported *afterwards*. My misconception in https://github.com/pyinstaller/pyinstaller/issues/1592#issuecomment-147230772 was: I thought these names already effect imports within the currently process module - which is true, or maybe true, in a single case.

This single case is `from . import abc` (with relative_import being `.`). In this single case, `abc` ought to be a submodule, otherwise this import statement would be crap.

So we can safely ignore this corner-case. Esp. since the only consequence would be a wrong entry in the `warnlog` file.

(It took me a punch of hours to verify this. So I added some comments to `modulegraph` describing it, see 0d67c6a284ce3ca28506524c533a2801f5465e02)

**Re 1:**

In PyInstaller 2.1, `attrs` added identifiers to the module's namespace so they exist when imported. But a) modulegraph does never look at these attributes b) has it's own mechanism for looking them up: `globalnames`. And c) `globalnames` is much better since it uses the names actually used in the module.

**Re 3:**

I propose removing all `attrs` from the hooks and eventually remove the hooks at all if they only set `attrs`. 

**Any opinions?**
